orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go

```diff
@@ -0,0 +1,170 @@
+
+type Store interface {
+	SetTokenInfo(tokenInfo protos.TokenInfo) error
+	GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error)
+	GetTokenInfoFromNonce(nonce string) (*protos.TokenInfo, error)
+}
+
+type blobstoreStore struct {
+	factory blobstore.StoreFactory
+}
+
+func NewBlobstoreStore(factory blobstore.StoreFactory) Store {
+	return &blobstoreStore{factory: factory}
+}
+
+func (b *blobstoreStore) SetTokenInfo(tokenInfo protos.TokenInfo) error {
+	networkID := tokenInfo.GatewayDeviceInfo.NetworkId
+	logicalID := tokenInfo.GatewayDeviceInfo.LogicalId
+
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return err
+	}
+	defer store.Rollback()
+
+	isUnique, err := b.isNonceUnique(store, tokenInfo.Nonce)
+	if err != nil {
+		return err
+	}
+	if !isUnique {
+		return errors.Errorf("token is not unique")
+	}
+
+	// Write to 2 blobstores so that we can key for tokenInfo with both the logicalID and the nonce
+
+	logicalIDBlob, err := tokenInfoToBlob(bootstrapper.LogicalIDToTokenInfo, logicalID, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(networkID, blobstore.Blobs{logicalIDBlob})
+	if err != nil {
+		return err
+	}
+
+	nonceBlob, err := tokenInfoToBlob(bootstrapper.NonceTokenToInfoMap, tokenInfo.Nonce, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(placeholderNetworkID, blobstore.Blobs{nonceBlob})
+	if err != nil {
+		return err
+	}
+
+	return store.Commit()
+}
+
+func (b *blobstoreStore) GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error) {
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer store.Rollback()
+
+	logicalIDTK := storage.TK{
+		Type: string(bootstrapper.LogicalIDToTokenInfo),
+		Key:  logicalID,
+	}
+	logicalIDBlob, err := store.Get(networkID, logicalIDTK)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := tokenInfoFromBlob(logicalIDBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	return tokenInfo, store.Commit()
+}
+
+func (b *blobstoreStore) GetTokenInfoFromNonce(nonce string) (*protos.TokenInfo, error) {
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer store.Rollback()
+
+	nonceTK := storage.TK{
+		Type: string(bootstrapper.NonceTokenToInfoMap),
+		Key:  nonce,
+	}
+	nonceBlob, err := store.Get(placeholderNetworkID, nonceTK)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := tokenInfoFromBlob(nonceBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	return tokenInfo, store.Commit()
+}
+
+// isNonceUnique must be run within a transaction
+func (b *blobstoreStore) isNonceUnique(store blobstore.Store, nonce string) (bool, error) {
+	nonceTK := storage.TK{
+		Type: string(bootstrapper.NonceTokenToInfoMap),
+		Key:  nonce,
+	}
+	nonceBlob, err := store.Get(placeholderNetworkID, nonceTK)
+	if err != nil {
+		if err == merrors.ErrNotFound {
+			return true, nil
+		}
+		return false, err
+	}
```

Comment:
Nit for uniformity and reducing cyclomatic complexity, don't nest the err checking. Eg https://github.com/magma/magma/blob/4a29306de8048c40749a6f59135a018ef1d7a2ce/lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers.go#L294-L299